rclpy/rclpy/client.py

```diff
@@ -126,18 +112,46 @@ def call_async(self, request: SrvTypeRequest) -> Future:
         if not isinstance(request, self.srv_type.Request):
             raise TypeError()
 
-        with self.handle:
-            sequence_number = self.__client.send_request(request)
-        if sequence_number in self._pending_requests:
-            raise RuntimeError('Sequence (%r) conflicts with pending request' % sequence_number)
+        with self._lock:
+            with self.handle:
+                sequence_number = self.__client.send_request(request)
+            if sequence_number in self._pending_requests:
+                raise RuntimeError(f'Sequence ({sequence_number}) conflicts with pending request')
 
-        future = Future()
-        self._pending_requests[sequence_number] = future
+            future = Future()
+            self._pending_requests[sequence_number] = future
 
-        future.add_done_callback(self.remove_pending_request)
+            future.add_done_callback(self.remove_pending_request)
 
         return future
 
+    def get_pending_request(self, sequence_number: int) -> Future:
+        """
+        Get a future from the list of pending requests.
+
+        :param sequence_number: Number identifying the pending request.
+        :return: The future corresponding to the sequence_number.
+        :raises: KeyError if the sequence_number is not in the pending requests.
+        """
+        with self._lock:
+            return self._pending_requests[sequence_number]
+
+    def remove_pending_request(self, future: Future) -> None:
+        """
+        Remove a future from the list of pending requests.
+
+        This prevents a future from receiving a response and executing its done callbacks.
+
+        :param future: A future returned from :meth:`call_async`
+        """
+        for seq, req_future in self._pending_requests.items():
```

Comment:
If we lock the whole thing I think we can drop the `try ... except KeyError` from the method?

Reply:
Yeah I think so :+1: . I suppose the try-except is there because of the race.

Reply:
Done.

Reply:
> If we lock the whole thing I think we can drop the try ... except KeyError from the method?
> Yeah I think so +1 . I suppose the try-except is there because of the race.

I don't think that was releated to the race.
The remove_pending_requests() method is used to avoid leaking when the server never answers and you're not interested in the response anymore (e.g. after waiting with a timeout).
The request might have been removed from the `_pending_requests` list before the user called the method, e.g. because a response just arrived between the timeout expired and the method call.

The try...except is needed, even with the lock.

Reply:
ah yes, I misread the code, thanks!